http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/SslContextProvider.java

```diff
@@ -0,0 +1,100 @@
+import javax.net.ssl.TrustManagerFactory;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.http.Protocol;
+import software.amazon.awssdk.http.TlsTrustManagersProvider;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+@SdkInternalApi
+public final class SslContextProvider {
+    private static final Logger log = Logger.loggerFor(SslContextProvider.class);
+    private final Protocol protocol;
+    private final SslProvider sslProvider;
+    private final TrustManagerFactory trustManagerFactory;
+    private final KeyManagerFactory keyManagerFactory;
+
+    public SslContextProvider(NettyConfiguration configuration, Protocol protocol, SslProvider sslProvider) {
+        this.protocol = protocol;
+        this.sslProvider = sslProvider;
+        this.trustManagerFactory = getTrustManager(configuration);
+        this.keyManagerFactory = getKeyManager(configuration);
+    }
+
+    public SslContext sslContext() {
+        try {
+            return SslContextBuilder.forClient()
+                                    .sslProvider(sslProvider)
+                                    .ciphers(getCiphers(), SupportedCipherSuiteFilter.INSTANCE)
+                                    .trustManager(trustManagerFactory)
+                                    .keyManager(keyManagerFactory)
+                                    .build();
+        } catch (SSLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private List<String> getCiphers() {
+        // HTTP/2: per Rfc7540, there is a blocked list of cipher suites for HTTP/2, so setting
+        // the recommended cipher suites directly here
+        // See https://tools.ietf.org/html/rfc7540#section-9.2.2
+        // HTTP/1.1: return null so that the default ciphers suites will be used
+        // https://github.com/netty/netty/blob/0dc246eb129796313b58c1dbdd674aa289f72cad/handler/src/main/java/io/netty/handler
+        // /ssl/SslUtils.java
+        return protocol.equals(Protocol.HTTP2) ? Http2SecurityUtil.CIPHERS : null;
+    }
+
+    private TrustManagerFactory getTrustManager(NettyConfiguration configuration) {
+        TlsTrustManagersProvider tlsTrustManagersProvider = configuration.tlsTrustManagersProvider();
+        Validate.isTrue(tlsTrustManagersProvider == null || !configuration.trustAllCertificates(),
+                        "A TlsTrustManagerProvider can't be provided if TrustAllCertificates is also set");
+
+        if (tlsTrustManagersProvider != null) {
+            return StaticTrustManagerFactory.create(tlsTrustManagersProvider.trustManagers());
+        }
+
+        if (configuration.trustAllCertificates()) {
+            log.warn(() -> "SSL Certificate verification is disabled. This is not a safe setting and should only be "
+                           + "used for testing.");
+            return InsecureTrustManagerFactory.INSTANCE;
+        }
+
+        return null;
```

Comment:
This return statement shouldn't get executed. Should we throw an exception or restructure the function?

Reply:
It actually will get executed in most cases. Passing null means the system default trust manager will be used. I can add a comment there to make it more clear

https://github.com/netty/netty/blob/0dc246eb129796313b58c1dbdd674aa289f72cad/handler/src/main/java/io/netty/handler/ssl/SslContextBuilder.java#L264-L270

Reply:
Got it, I misread the validate statement.